### PR TITLE
Normalize the URL in the `baseUrl` option

### LINKED ISSUE #562: Document the behavior of the `baseUrl` option
in common senarios, we used to deploy our api with a url `{host}/api`, so we may use `got` like this:

```js
const client = got.extend({
  baseUrl: 'http://xxx.com/api/v1',
});

client.get('/stuffs')
```

somehow, cuz the `baseUrl` just support a [WHATWG URL](https://nodejs.org/api/url.html#url_class_url) currently, we have to add a `/api/v1` prefix for each url and it really sucks for tons of long urls.

### LINKED ISSUE #574: Document the `response.retryCount` property
Looks like the [`response` documentation](https://github.com/sindresorhus/got#goturl-options) became outdated with the 9.0 release because the new [`retryCount` property](https://github.com/sindresorhus/got/blob/8f30f1fdefaa96ad55758a352730d00697862952/source/request-as-event-emitter.js#L51) isn't documented.

### PR DESCRIPTION
**Pull request by szmarczak**
Fixes #562

### CODE REVIEW

**Comment by sindresorhus** [comment]
The docs needs to be updated too.

**Comment by sindresorhus** [comment]
> Rewrite the `baseUrl` option

Would be more descriptive to title the PR:

> Normalize the URL in the `baseUrl` option

**Comment by sindresorhus** [comment]
Can you comment on why it was necessary to split it out into a preNormalize step?

File: source/normalize-arguments.js [hunk 1]
```diff
@@ -197,42 +251,7 @@ module.exports = (url, options, defaults) => {
...
-			}
-		}
-	} else {
-		throw new TypeError(`Parameter \`hooks\` must be an object, not ${is(options.hooks)}`);
-	}
-
 	return options;
 };
+
+module.exports.prenormalize = prenormalize;
```
[the remark this answers is not in the document]
Comment (szmarczak) [reply source/normalize-arguments.js]: Sure

File: test/arguments.js [hunk 2]
```diff
@@ -194,6 +198,18 @@ test('allows extra keys in `hooks`', async t => {
...
+	const instanceA = got.extend({baseUrl: `${s.url}/test`});
+	const {body} = await instanceA('/foobar');
+	t.is(body, `/test/foobar`);
+});
+
+test('backslash is optional (baseUrl)', async t => {
+	const instanceA = got.extend({baseUrl: `${s.url}/test/`});
+	const {body} = await instanceA('/foobar');
+	t.is(body, `/test/foobar`);
+});
```
[the remark this answers is not in the document]
Comment (szmarczak) [reply test/arguments.js]: yup

**Comment by szmarczak** [comment]
> Can you comment on why it was necessary to split it out into a preNormalize step?

Code comment or GitHub comment?

**Comment by sindresorhus** [comment]
Just a comment on this thread is fine.

File: source/normalize-arguments.js [hunk 3]
```diff
@@ -13,6 +13,47 @@ const knownHookEvents = require('./known-hook-events');
 
 const retryAfterStatusCodes = new Set([413, 429, 503]);
 
+const preNormalize = options => {
```
Comment (szmarczak) [inline source/normalize-arguments.js]: I've split `normalize-argument.js` into two parts:
- `preNormalize` handles all the stuff which is related with static options like `baseUrl`, `followRedirect`, `hooks` etc.
  It is used in `got.create()` to  normalize `defaults.options`.
 - `normalize` does `preNormalize` + handles all the stuff related with dynamic options like `url`, `headers`, `body` etc.

**Comment by sindresorhus** [comment]
Looking great!

68747470733a2f2f6d656469612e67697068792e636f6d2f6d656469612f6c3045786b3845557a534c7372457245512f67697068792e676966

**Pull request merged**